README.md
# Soccer Module for Joomla 4.x

Soccer results, tables and schedules for Joomla based on jbuli (Meister1900)

## Soccer schedules

Joomla module to display schedules of a team in various leagues (to be chosen in backend)

## Soccer results

Joomla module to display results of gamedays of various leagues (to be chosen in backend)

## Soccer tables

Joomla module to display current table of various leagues (to be chosen in backend)


mod_soccer_results/CHANGELOG.md
# Changelog

* 0.1 initial version


mod_soccer_results/install.php
<?php

// No direct access to this file
defined('_JEXEC') or die;

use Joomla\CMS\Language\Text;
use Joomla\CMS\Log\Log;
use Joomla\CMS\Factory;

class mod_soccer_resultsInstallerScript {
    /**
     * Extension script constructor.
     *
     * @return  void
     */
    public function __construct() {
        $this->minimumJoomla = '4.0';
    }    

    /**
     * Function called before extension installation/update/removal procedure commences
     *
     * @param   string            $type    The type of change (install, update or discover_install, not uninstall)
     * @param   InstallerAdapter  $parent  The class calling this method
     *
     * @return  boolean  True on success
     */
    function preflight($type, $parent) {
        // Check for the minimum Joomla version before continuing
        if (!empty($this->minimumJoomla) && version_compare(JVERSION, $this->minimumJoomla, '<')) {
            Log::add(Text::sprintf('JLIB_INSTALLER_MINIMUM_JOOMLA', $this->minimumJoomla), Log::WARNING, 'jerror');

            return false;
        }
        
		echo Text::_('MOD_SOCCER_RESULTS_INSTALLERSCRIPT_PREFLIGHT');
        
        echo $this->minimumJoomla;

        return true;
    }
	
    /**
     * Function called after extension installation/update/removal procedure commences
     *
     * @param   string            $type    The type of change (install, update or discover_install, not uninstall)
     * @param   InstallerAdapter  $parent  The class calling this method
     *
     * @return  boolean  True on success
     */
    function postflight($type, $parent) {
        return true;
    }

    /**
     * Method to install the extension
     *
     * @param   InstallerAdapter  $parent  The class calling this method
     *
     * @return  boolean  True on success
     */
    function install($parent) {
        $this->setupDatabase();
        return true;
    }    

    /**
     * Method to update the extension
     *
     * @param   InstallerAdapter  $parent  The class calling this method
     *
     * @return  boolean  True on success
     */
    function update($parent) {
        $this->setupDatabase();
        return true;
    }    

    /**
     * Method to uninstall the extension
     *
     * @param   InstallerAdapter  $parent  The class calling this method
     *
     * @return  boolean  True on success
     */
    function uninstall($parent) {
        // $db = Factory::getDbo();
		$db = Factory::getContainer()->get('DatabaseDriver');
        $query = 'DROP TABLE '.$db->quoteName('#__soccer_results');

        $db->setQuery($query);
        $db->execute();

        return true;
    }

    private function setupDatabase() {
        // $db = Factory::getDbo();
		$db = Factory::getContainer()->get('DatabaseDriver');
        $query = 'CREATE TABLE IF NOT EXISTS '.$db->quoteName('#__soccer_results').' (ID INT NOT NULL AUTO_INCREMENT PRIMARY KEY, bezeichnung_webservice VARCHAR(100), bezeichnung_kurz VARCHAR(100), bezeichnung_mittel VARCHAR(100), dateiname_logo VARCHAR(100))';

        $db->setQuery($query);
        $db->execute();

        $query = 'TRUNCATE TABLE '.$db->quoteName('#__soccer_results');
        $db->setQuery($query);
        $db->execute();

        $query = "INSERT INTO ".$db->quoteName('#__soccer_results')." VALUES
				(1, 'VfL Wolfsburg', 'WOL', 'Wolfsburg', 'wolfsburg.png'), 
				(2, 'FC Schalke 04', 'S04', 'Schalke', 'schalke.png'),
				(3, 'TSG 1899 Hoffenheim', 'HOF', 'Hoffenheim', 'hoffenheim.png'), 
				(4, 'FC Bayern München', 'FCB', 'Bayern', 'bayern.png'),
				(5, 'Werder Bremen', 'BRE', 'Bremen', 'bremen.png'),
				(6, 'Borussia Mönchengladbach', 'BMG', 'Gladbach', 'gladbach.png'), 
				(7, 'Eintracht Frankfurt', 'FRA', 'Frankfurt', 'frankfurt.png'), 
				(8, '1. FSV Mainz 05', 'MAI', 'Mainz', 'mainz.png'),
				(9, 'SC Freiburg', 'FRE', 'Freiburg', 'freiburg.png'), 
				(10, 'Hamburger SV', 'HSV', 'Hamburg', 'hamburg.png'), 
				(11, 'Hannover 96', 'HAN', 'Hannover', 'hannover.png'), 
				(12, 'Borussia Dortmund', 'BVB', 'Dortmund', 'dortmund.png'), 
				(13, 'VfB Stuttgart', 'STU', 'Stuttgart', 'stuttgart.png'),
				(14, 'Bayer 04 Leverkusen', 'LEV', 'Leverkusen', 'leverkusen.png'), 
				(15, 'FC Augsburg', 'AUG', 'Augsburg', 'augsburg.png'),
				(16, 'Hertha BSC', 'BSC', 'Hertha', 'hertha.png'),
				(17, 'SC Paderborn 07', 'PBO', 'Paderborn', 'paderborn.png'), 
				(18, '1. FC Köln', 'KLN', 'Köln', 'koeln.png'),
				(19, 'Fortuna Düsseldorf', 'DÜS', 'Düsseldorf', 'duesseldorf.png'), 
				(20, 'Eintracht Braunschweig', 'BRA', 'Braunschw.', 'braunschweig.png'),
				(21, 'RB Leipzig', 'LPZ', 'Leipzig', 'leipzig.png'), 
				(22, 'VfR Aalen', 'AAL', 'Aalen', 'aalen.png'),
				(23, 'FC St. Pauli', 'STP', 'St. Pauli', 'pauli.png'),
				(24, 'FC Ingolstadt 04', 'ING', 'Ingolstadt', 'ingolstadt.png'), 
				(25, 'VfL Bochum', 'BOC', 'Bochum', 'bochum.png'), 
				(26, 'SpVgg Greuther Fürth', 'FÜR', 'Fürth', 'fuerth.png'),
				(27, '1. FC Heidenheim 1846', 'HEI', 'Heidenh.', 'heidenheim.png'), 
				(28, 'FSV Frankfurt', 'FRA', 'Frankfurt', 'fsvfrankfurt.png'), 
				(29, '1. FC Nürnberg', 'NÜR', 'Nürnberg', 'nuernberg.png'), 
				(30, 'SV Darmstadt 98', 'DAR', 'Darmstadt', 'darmstadt.png'), 
				(31, 'SV Sandhausen', 'SAN', 'Sandhausen', 'sandhausen.png'),
				(32, 'Karlsruher SC', 'KSC', 'Karlsruhe', 'karlsruhe.png'), 
				(33, '1. FC Union Berlin', 'BER', 'Berlin', 'berlin.png'),
				(34, '1. FC Kaiserslautern', 'FCK', 'Kaisersl.', 'kaiserslautern.png'),
				(35, 'TSV 1860 München', 'MÜN', 'Müchen', 'muenchen.png'), 
				(36, 'Erzgebirge Aue', 'AUE', 'Aue', 'aue.png'),
				(37, 'MSV Duisburg', 'DUI', 'Duisburg', 'duisburg.png'),
				(38, 'Arminia Bielefeld', 'BIE', 'Bielefeld', 'bielefeld.png'),
				(39, 'Stoke City FC', 'STK', 'Stoke City', 'stoke.png'),
				(40, 'Manchester City FC', 'MCI', 'ManCity', 'city.png'),
				(41, 'Arsenal FC', 'ARS', 'Arsenal', 'arsenal.png'),
				(42, 'Sunderland AFC', 'SUN', 'Sunderland', 'sunderland.png'),
				(43, 'Manchester United FC', 'MUN', 'ManUnited', 'manu.png'),
				(44, 'West Ham United FC', 'WHU', 'Westham', 'westham.png'),
				(45, 'FC Southampton', 'SOU', 'Southampton', 'southampton.png'),
				(46, 'Swansea City', 'SWA', 'Swansea', 'swansea.png'),
				(47, 'Leicester City', 'LEI', 'Leicester', 'leicester.png'),
				(48, 'FC Watford', 'WAT', 'Watford', 'watford.png'),
				(49, 'Norwich City', 'NOR', 'Norwich', 'norwich.png'),
				(50, 'West Bromwich Albion', 'WBA', 'Bromwich', 'bromwich.png'),
				(51, 'Tottenham Hotspur FC', 'TOT', 'Tottenham', 'tottenham.png'),
				(52, 'Chelsea FC', 'CHE', 'Chelsea', 'chelsea.png'),
				(53, 'AFC Bournemouth', 'BOU', 'Bournem.', 'bournemouth.png'),
				(54, 'Newcastle United', 'NEW', 'Newcastle', 'newcastle.png'),
				(55, 'Liverpool FC', 'LIV', 'Liverpool', 'liverpool.png'),
				(56, 'Everton FC', 'EVE', 'Everton', 'everton.png'),
				(59, 'Crystal Palace', 'CRY', 'Crystal Palace', 'palace.png'),
				(58, 'Aston Villa FC', 'AVL', 'Aston Villa', 'villa.png'),
				(60, 'Juventus Turin', 'JUV', 'Juventus', 'juventus.png'),
				(61, 'FC Empoli', 'EMP', 'FC Empoli', 'empoli.png'),
				(62, 'Inter Mailand', 'INT', 'Mailand', 'mailand.png'),
				(63, 'Chievo Verona', 'CHI', 'Verona', 'verona.png'),
				(64, 'Lazio Rom', 'LAZ', 'Lazio Rom', 'lazio.png'),
				(65, 'Carpi FC', 'CAR', 'Carpi FC', 'carpi.png'),
				(66, 'AC Florenz', 'FLO', 'AC Florenz', 'florenz.png'),
				(67, 'US Palermo', 'PAL', 'US Palermo', 'palermo.png'),
				(68, 'Udinese Calcio', 'CAL', 'Udinese', 'udinese.png'),
				(69, 'FC Genua', 'GEN', 'FC Genua', 'genua.png'),
				(70, 'Sassuolo', 'SAS', 'Sassuolo', 'calcio.png'),
				(71, 'Sampdoria Genua', 'GEN', 'Genua', 'genua.png'),
				(72, 'FC Bologna', 'BOL', 'FC Bologna', 'bologna.png'),
				(73, 'AC Mailand', 'MIL', 'AC Mailand', 'milan.png'),
				(74, 'Atalanta Bergamo', 'BER', 'Atl. Bergamo', 'bergamo.png'),
				(75, 'Frosinone Calcio', 'FRO', 'Fros. Calcio', 'frosione_calcio.png'),
				(76, 'Hellas Verona', 'VER', 'Verona', 'hellas_verona.png'),
				(77, 'SSC Neapel', 'NEP', 'Neapel', 'neapel.png'),
				(78, 'AS Rom', 'ROM', 'AS Rom', 'rom.png'),
				(79, 'FC Turin', 'TUR', 'FC Turin', 'turin.png'),
				(80, 'FC Malaga', 'MAL', 'FC Malaga', 'malaga.png'),
				(81, 'FC Barcelona', 'FCB', 'Barcelona', 'barcelona.png'),
				(82, 'Celta Vigo', 'VIG', 'Celta Vigo', 'vigo.png'),
				(83, 'Atletico Madrid', 'ATM', 'Atletico', 'atletico.png'),
				(84, 'Rayo Vallecano', 'RAY', 'Vallecano', 'vallecano.png'),
				(85, 'Levante UD', 'LEV', 'Levante', 'levante.png'),
				(86, 'FC Sevilla', 'SEV', 'FC Sevilla', 'sevilla.png'),
				(87, 'Athletic Bilbao', 'BIL', 'Bilbao', 'bilbao.png'),
				(88, 'UD Las Palmas', 'PAL', 'Palmas', 'laspalmas.png'),
				(89, 'Granada CF', 'GRA', 'Granada CF', 'granada.png'),
				(90, 'Sporting Gijon', 'GIJ', 'Gijon', 'gijon.png'),
				(91, 'FC Getafe', 'GET', 'Getafe', 'getafe.png'),
				(92, 'Real Sociedad', 'SOC', 'Sociedad', 'sociedad.png'),
				(93, 'Deportivo La Coruna', 'DLC', 'Deportivo', 'lacoruna.png'),
				(94, 'SD Eibar', 'EIB', 'SD Eibar', 'eibar.png'),
				(95, 'FC Valencia', 'VAL', 'Valencia', 'valencia.png'),
				(96, 'FC Villareal', 'VIL', 'Villareal', 'villareal.png'),
				(97, 'Betis Sevilla', 'BET', 'Betis Sevilla', 'sevilla.png'),
				(98, 'Real Madrid', 'RMA', 'Real', 'real.png'),
				(99, 'Espanyol Barcelona', 'ESP', 'Espanyol', 'esp_barcelona.png'),
				(100, 'Schachtjor Donezk', 'DON', 'Donezk', 'donezk.png'),
				(101, 'Paris St. Germain', 'PSG', 'Paris', 'paris.png'),
				(102, 'Malmö FF', 'MAL', 'Malmö FF', 'malmö.png'),
				(103, 'Benfica Lissabon', 'LIS', 'Lissabon', 'lissabon.png'),
				(104, 'KAA Gent', 'GEN', 'KAA Gent', 'gent.png'),
				(105, 'Zenit St. Petersburg', 'PET', 'Petersburg', 'zenit.png'),
				(106, 'VFL Wolfsburg', 'WOL', 'VFL Wolfsburg', 'wolfsburg.png'),
				(107, 'FC Arsenal', 'ARS', 'Arsenal', 'arsenal.png'),
				(108, 'PSV Eindhoven', 'PSV', 'Eindhoven', 'eindhoven.png'),
				(109, 'Dynamo Kiew', 'KIV', 'Kiew', 'kiew.png'),
				(110, 'Manchester City', 'MCI', 'ManCity', 'city.png'),
				(111, 'Olympique Lyon', 'LYO', 'Lyon', 'lyon.png'),
				(112, 'FC Porto', 'POR', 'Porto', 'porto.png'),
				(113, 'Maccabi Tel Aviv', 'AVI', 'Tel Aviv', 'telaviv.png'),
				(114, 'Olympiakos Piräus', 'PIR', 'Piräus', 'piräus.png'),
				(115, 'Dinamo Zagreb', 'ZAG', 'Zagreb', 'zagreb.png'),
				(116, 'FC BATE Borisov', 'BOR', 'Borisov', 'borisov.png'),
				(117, 'FK Astana', 'AST', 'FK Astana', 'astana.png'),
				(118, 'Galatasaray Istanbul', 'GAL', 'Istanbul', 'galatasaray.png'),
				(119, 'ZSKA Moskau', 'MOS', 'Moskau', 'moskau.png'),
				(120, 'Manchester United', 'MUN', 'Manchester', 'manu.png'),
				(121, 'SG Dynamo Dresden', 'DRE', 'Dresden', 'dresden.png'),
				(122, 'FC Crotone', 'CRO', 'Crotone', 'crotone.png'),
				(123, 'Cagliari Calcio', 'CAG', 'Cagliari', 'cagliari.png'),
				(124, 'Delfino Pescara 1936', 'PES', 'Pescara', 'pescara.png'),
				(125, 'CA Osasuna', 'OSA', 'Osasuna', 'osasuna.png'),
				(126, 'Real Betis', 'BET', 'Betis', 'betis.png'),
				(127, 'Deportivo Alaves', 'ALA', 'Alaves', 'alaves.png'),
				(128, 'CD Leganes', 'LEG', 'Leganes', 'leganes.png'),
				(129, 'Hull City AFC', 'HUL', 'Hull', 'hull.png'),
				(130, 'Burnley FC', 'BUR', 'Burnley', 'burnley.png'),
				(131, 'FC Middlesbrough', 'MID', 'Middlesbrough', 'middlesbrough.png'),
				(132, 'Würzburger Kickers', 'WÜB', 'Würzburg', 'wuerzburg.png'),
				(133, 'FC Basel', 'BAS', 'Basel', 'basel.png'),
				(134, 'Ludogorez Rasgrad', 'RAS', 'Rasgrad', 'rasgrad.png'),
				(135, 'Besiktas Istanbul', 'IST', 'Istanbul', 'besiktas.png'),
				(136, 'Celtic Glasgow', 'GLA', 'Glasgow', 'glasgow.png'),
				(137, 'FK Rostow', 'ROS', 'Rostow', 'rostow.png'),
				(138, 'AS Monaco', 'MON', 'Monaco', 'monaco.png'),
				(139, 'Legia Warschau', 'WAR', 'Warschau', 'warschau.png'),
				(140, 'Sporting Lissabon', 'LIS', 'Lissabon', 'lissabon.png'),
				(141, 'FC Brügge', 'BRÜ', 'Brügge', 'bruegge.png'),
				(142, 'FC Kopenhagen', 'KOP', 'Kopenhagen', 'kopenhagen.png'),
				(143, 'Holstein Kiel', 'KIE', 'Kiel', 'kiel.png'),
                (144, 'Jahn Regensburg', 'REG', 'Regensburg', 'regensburg.png'),
				(145, 'Leverkusen', 'LEV', 'Leverkusen', 'leverkusen.png'),
				(146, 'FC Bayern', 'FCB', 'Bayern', 'bayern.png'),
                (147, 'Bayer Leverkusen', 'LEV', 'Leverkusen', 'leverkusen.png'),
                (148, '1. FC Magdeburg', 'MAG', 'Magdeburg', 'magdeburg.png'),
			    (149, 'VfL Osnabrück', 'OSN', 'Osnabrück', 'osnabrueck.png'),
			    (150, 'SV Wehen Wiesbaden', 'WIS', 'Wiesbaden', 'wiesbaden.png'),
			    (151, 'Karlsruher SC', 'KSC', 'Karlsruhe', 'karlsruhe.png'),
				(152, 'FC Hansa Rostock', 'ROS', 'Rostock', 'rostock.png');
			   ";

        $db->setQuery($query);
        $db->execute();

        $cachefile = JPATH_BASE."/../modules/mod_soccer_results/cache.txt";
        if (is_readable($cachefile)) {
            unlink($cachefile);
        }
    }
}


mod_soccer_results/helper.php
<?php
/**
 * helper.php 
 */

use Joomla\CMS\Log\Log;
use Joomla\CMS\Factory;

class modSoccerResultsHelper
{
    /**
     * Constructor
     */
    public function __construct($module)
    {
        // Load Bootstrap and JQuery
        
        JHtml::_('bootstrap.framework');
        JHtml::_('jquery.framework'); 

        $app = Factory::getApplication();
        $document = $app->getDocument();        
        $document->addScriptDeclaration('
      jQuery(document).ready(function() {
        
        change_spieltag_' . $module->id . '();
        jQuery(document).on("change", "#spielplan_' . $module->id . '", change_spieltag_' . $module->id . ');
      });
        
      function change_spieltag_' . $module->id . '() {
        jQuery("#soccer_results_loading_' . $module->id . '").show();
        jQuery.post( "' . JURI::base() . 'index.php",
            {
              option: "com_ajax",
              module: "soccer_results",
              Itemid: "' . $app->getMenu()->getActive()->id . '",
              method: "get",
              format: "json",
              titel: "' . $module->title . '",
              spieltag: jQuery("#spieltag_' . $module->id . ' option:selected").text(),
            },
            function(data){
              jQuery("#soccer_results_loading_' . $module->id . '").hide();
              if (data.success == false) {
                jQuery("#spielplan_' . $module->id . '").html(data.message);
              } else {
                jQuery("#spielplan_' . $module->id . '").html(data.data);
                jQuery(".hasTooltip").tooltip({html: "true"});
              }
            }
        ).fail(function(xhr) {
		  try {
			// Ungewollten Output von anderen Plugins wie GoogleAnalytics oder PHP Meldungen wegschneiden
			data = jQuery.parseJSON(xhr.responseText.substring(xhr.responseText.indexOf("success")-2));
		  }
		  catch (e) {
			alert("Fehlerhafter JSON Response - Doku pruefen!");
		  };
	      jQuery("#soccer_results_loading_' . $module->id . '").hide();
          if (data.success == false) {
            jQuery("#spielplan_' . $module->id . '").html(data.message);
          } else {
            jQuery("#spielplan_' . $module->id . '").html(data.data);
            jQuery(".hasTooltip").tooltip({html: "true"});
          }
       });
      };
    ');
    }

    /**
     * fetch data from api using curl or file_get_contents
     */
    public static function fetchdata($url, $timeout)
    {
        Log::add('in fetchdata ', Log::ERROR, 'my-debug-log');
        if (function_exists('curl_version')) {
            $curl = curl_init();
            curl_setopt($curl, CURLOPT_URL, $url);
            curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
            curl_setopt($curl, CURLOPT_TIMEOUT, $timeout);
            $content = curl_exec($curl);
            curl_close($curl);
            return $content;
        } elseif (ini_get('allow_url_fopen')) {
            $context = stream_context_create([
        'http' => [ 'timeout' => $timeout ]
      ]);

            return file_get_contents($url, 0, $context);
        } else {
            return false;
        }
    }

    /**
     * AJAX Endpoint
     */
    public static function getAjax()
    {
        $jinput = Factory::getApplication()->input;
        $module = JModuleHelper::getModule('soccer_results', $jinput->get('titel', 'default_value', 'filter'));
        $db = Factory::getContainer()->get('DatabaseDriver');
        // $db = Factory::getDbo();

        $jparams = new JRegistry();
        $jparams->loadString($module->params);

        $tage = ['So.', 'Mo.', 'Di.', 'Mi.', 'Do.', 'Fr.', 'Sa.'];
        $clrunden = ['', 'Gruppe A', 'Gruppe B', 'Gruppe C', 'Gruppe D', 'Gruppe E', 'Gruppe F', 'Gruppe G', 'Gruppe H', 'Achtelfinale', 'Viertelfinale', 'Halbfinale', 'Finale'];

        // Teams aus der Joomla Tabelle holen
        $query = 'SELECT '.$db->quoteName('bezeichnung_webservice').', '.$db->quoteName('bezeichnung_kurz').', '.$db->quoteName('bezeichnung_mittel').', '.$db->quoteName('dateiname_logo').' FROM '.$db->quoteName('#__soccer_results');
        $db->setQuery($query);
        $teams = $db->loadAssocList('bezeichnung_webservice');
        $liga = $jparams->get('league');

        // Spieltag ermitteln
        if ($jinput->get('spieltag', 'default_value', 'filter') != '') {
            $spieltag = $jinput->get('spieltag', 'default_value', 'filter');
            if (! is_numeric($spieltag)) {
                $spieltag = array_search($spieltag, $clrunden);
            }
        } elseif ($jparams->get('matchday') != 0 && $jparams->get('matchday') != -1) {
            $spieltag = $jparams->get('matchday');
        } else {
            $spieltag = self::fetchdata('https://www.openligadb.de/api/getcurrentgroup/' .$liga, $jparams->get('timeout'));

            if ($spieltag === false) {
                // Kein Spieltag vom Webservice -> den vom letzten Mal nehmen
                if ($jparams->get('lastCurrentMatchday') != '') {
                    $spieltag = $jparams->get('lastCurrentMatchday');
                } else {
                    $spieltag = 1;
                }
            } else {
                $spieltag = json_decode($spieltag);
                $spieltagsname = $spieltag->groupName;
                $spieltag = $spieltag->groupOrderID;

                if ($liga == 'cl1617') {
                    $spieltag = array_search($spieltagsname, $clrunden);
                    if (! $spieltag || $spieltag < 9) {
                        $spieltag = 1;
                    }
                }

                // Aktueller Spieltag hat sich geändert -> Als Parameter lastCurrentMatchday speichern
                if ($jparams->get('lastCurrentMatchday') != $spieltag) {
                    $jparams->set('lastCurrentMatchday', $spieltag);
                    $module->params = $jparams->toString();
                    $dbtable = JTable::getInstance('module');
                    $dbtable->save((array)$module);
                }
            }
        }

        // Wenn -1 eingestellt dann vorherigen Spieltag anzeigen
        if ($jparams->get('matchday') == '-1' && $spieltag > 1) {
            $spieltag -= 1;
        }
        $saison = $jparams->get('season');
        
        // Cache lesen
        $cachefile = JPATH_BASE."/modules/mod_soccer_results/cache.txt";
        if (is_readable($cachefile)) {
            $cache = file_get_contents($cachefile);
        } else {
            $cache = self::fetchdata('https://www.fcb-lippach.de/modules/mod_soccer_results/cache.txt', 10);

            if ($cache != false) {
                file_put_contents($cachefile, $cache);
            }
        }
        $paarungen_cache = unserialize($cache);

        // Letzte Änderung ermitteln
        $lastchange = self::fetchdata('https://api.openligadb.de/getlastchangedate/' . $liga . '/' . $saison . '/' . $spieltag, $jparams->get('timeout'));
        
        if ($lastchange === false) {
            // Kein Datum vom Webservice -> Datum aus dem Cache holen
            if ($paarungen_cache[$spieltag . $liga . $saison]) {
                $lastchange = array_keys($paarungen_cache[$spieltag . $liga . $saison]);
                $lastchange = $lastchange[0];
            }
        } else {
            $lastchange = strtotime(json_decode($lastchange));
        }
        
        // Spieltag mit diesem Stand schon im Cache?
        if (isset($paarungen_cache[$spieltag . $liga . $saison][$lastchange])) {
            $paarungen = $paarungen_cache[$spieltag . $liga . $saison][$lastchange];
        } else {
            // Daten abrufen und in den Cache schreiben
            $paarungen = self::fetchdata('https://api.openligadb.de/getmatchdata/' . $liga . '/' . $saison . '/' . $spieltag, $jparams->get('timeout'));

            if ($paarungen === false || stristr($paarungen, 'Maximale Abfrageanzahl von 1000 Abfragen pro Tag erreicht!') != false) {
                // Webservice nicht erreichbar, prüfen ob Spieltag mit älterem Stand im Cache ist
                if (is_array($paarungen_cache[$spieltag . $liga . $saison])) {
                    $paarungen = reset($paarungen_cache[$spieltag . $liga . $saison]);
                }
            } else {
                $paarungen = json_decode($paarungen);
                unset($paarungen_cache[$spieltag . $liga . $saison]);
                $paarungen_cache[$spieltag . $liga . $saison][$lastchange] = $paarungen;
                file_put_contents($cachefile, serialize($paarungen_cache));
            }
        }

        // Prüfen wie viele Ergebnisse zu diesem Spieltag vorliegen
        $anzahl_ergebnisse = 0;
        $anzahl_live = 0;
        if ($paarungen) {
            $bezeichnung = 'bezeichnung_mittel';
            foreach ($paarungen as $partie) {
                if (isset($partie->matchResults[0])) {
                    $ergebnisse = $partie->matchResults[0];
                    if ($ergebnisse instanceof stdClass) {
                        $anzahl_ergebnisse++;
                        if ($jparams->get('longnames') == '1') {
                            $bezeichnung = 'bezeichnung_mittel';
                        } else {
                            $bezeichnung = 'bezeichnung_kurz';
                        }

                        if ($partie->matchIsFinished == false) {
                            $anzahl_live++;
                        }
                    }
                }
            }
        }

        // Start HTML OUTPUT
        $table = "<table border='0' cellpadding='1' cellspacing='1'>\r\n";

        // Spieltag Dropdown
        if ($liga == 'cl1617') {
            $breite = '100';
        } else {
            $breite = '50';
        }
        $table .= "<tr>\r\n<td align='left' valign='middle' colspan='7' style='padding-bottom:10px;'><nobr>Spieltag:&nbsp;<select style='min-width: " . $breite . "px;' id='spieltag_" . $module->id . "'>";

        if ($liga == 'pl' || $liga == 'sa' || $liga == 'pd') {
            $spieltage = 38;
        } elseif ($liga == 'cl1617') {
            $spieltage = 12;
        } else {
            $spieltage = 34;
        }

        for ($i=1;$i<=$spieltage;$i++) {
            if ($liga == 'cl1617') {
                $anzeige = $clrunden[$i];
            } else {
                $anzeige = $i;
            }

            if ($i == $spieltag && $i == $jparams->get('lastCurrentMatchday')) {
                $table .= "<option value='$i' style='font-weight:bold;' selected='selected'>$anzeige</option>";
            } elseif ($i == $spieltag) {
                $table .= "<option value='$i' selected='selected'>$anzeige</option>";
            } elseif ($i == $jparams->get('lastCurrentMatchday')) {
                $table .= "<option value='$i' style='font-weight:bold;'>$anzeige</option>";
            } else {
                $table .= "<option value='$i'>$anzeige</option>";
            }
        }

        $table .= "</select>&nbsp;&nbsp;&nbsp;<img id='soccer_results_loading_" . $module->id . "' src='".JURI::root()."modules/mod_soccer_results/images/ajax-loader.gif' style='display:none;'>";
        $table .= "</nobr></td>\r\n</tr>\r\n";

        // Live Spiele anzeigen
        if ($anzahl_ergebnisse > 0 && $anzahl_live == 1) {
            $table .= "<tr>\r\n<td align='center' valign='middle' colspan='7'><font color='red'><b>JETZT EIN SPIEL LIVE!</b></font></td>\r\n</tr>\r\n";
        } elseif ($anzahl_ergebnisse > 0 && $anzahl_live > 1) {
            $table .= "<tr>\r\n<td align='center' valign='middle' colspan='7'><font color='red'><b>JETZT ".$anzahl_live." SPIELE LIVE!</b></font></td>\r\n</tr>\r\n";
        }

        if (!$paarungen) {
            $table .= '</table>' . $jparams->get('timeout_error');

            return $table;
        }

        $i = 0;
        $termin = '';
        foreach ($paarungen as $partie) {
            $i++;

            if (trim($partie->team1->teamName) == $jparams->get('meinVerein') ||  trim($partie->team2->teamName) == $jparams->get('meinVerein')) {
                $style = $jparams->get('meinVereinCSS');
            } else {
                $style = '';
            }

            if ($termin != $partie->matchDateTime && !$jparams->get('kompakt')) {
                if ($i==1) {
                    $table .= "<tr>\r\n<td align='left' valign='middle' colspan='7'><b><i>".$tage[date("w", strtotime($partie->matchDateTime))]." ".date("d.m. H:i", strtotime($partie->matchDateTime))." Uhr</i></b></td>\r\n</tr>\r\n";
                } else {
                    $table .= "<tr>\r\n<td align='left' valign='middle' colspan='7' style='padding-top:10px;'><b><i>".$tage[date("w", strtotime($partie->matchDateTime))]." ".date("d.m. H:i", strtotime($partie->matchDateTime))." Uhr</i></b></td>\r\n</tr>\r\n";
                }
            }

            $table .= "<tr style='$style'>\r\n";

            if ($jparams->get('kompakt')) {
                if ($liga == 'cl1617') {
                    $table .= "<td align='left' valign='middle'>".date("d.m.", strtotime($partie->matchDateTime))."</td>\r\n";
                } else {
                    $table .= "<td align='left' valign='middle'>".$tage[date("w", strtotime($partie->matchDateTime))]."</td>\r\n";
                }
            }

            $termin = $partie->matchDateTime;

            // Team 1
            $table .= "<td align='left' valign='middle'><img style='width:20px; height:20px;' title='".$teams[trim($partie->team1->teamName)]['bezeichnung_mittel']."' alt='".$teams[trim($partie->team1->teamName)]['bezeichnung_mittel']."' src='".JURI::root()."modules/mod_soccer_results/images/".$teams[trim($partie->team1->teamName)]['dateiname_logo']."' /></td>\r\n";
            $table .= "<td align='left' valign='middle'>".$teams[trim($partie->team1->teamName)][$bezeichnung]."</td>\r\n";

            $table .= "<td align='left' valign='middle'>-&nbsp;</td>\r\n";

            // Team 2
            $table .= "<td align='left' valign='middle'><img style='width:20px; height:20px;' title='".$teams[trim($partie->team2->teamName)]['bezeichnung_mittel']."' alt='".$teams[trim($partie->team2->teamName)]['bezeichnung_mittel']."' src='".JURI::root()."modules/mod_soccer_results/images/".$teams[trim($partie->team2->teamName)]['dateiname_logo']."' /></td>\r\n";
            $table .= "<td align='left' valign='middle'>".$teams[trim($partie->team2->teamName)][$bezeichnung]."</td>\r\n";

            $tootip_text = "";
            $endergebnis = "";
            $halbzeitergebnis = "";
            if ($anzahl_ergebnisse > 0) {
                $table .= "<td align='left' valign='middle'>";
                $alle_ergebnisse = $partie->matchResults;
                if (isset($alle_ergebnisse[0])) {
                    if (!$partie->matchIsFinished && $alle_ergebnisse[0] instanceof stdClass) {
                        $tootip_text .= "<font color=red>";
                    }
                }
                $table .= "<nobr>&nbsp;";

                if (! is_array($alle_ergebnisse) || count($alle_ergebnisse) == 0) {
                    $tootip_text .= '-:- (-:-)';
                } else {
                    // Halbzeitergebnis / Endergebnis ermitteln
                    foreach ($alle_ergebnisse as $ergebnis) {
                        if ($ergebnis->resultName == 'Endergebnis') {
                            $endergebnis = $ergebnis->pointsTeam1.":".$ergebnis->pointsTeam2;
                        } elseif ($ergebnis->resultName == 'Halbzeitergebnis' || $ergebnis->resultName == 'Halbzeit') {
                            $halbzeitergebnis = " (".$ergebnis->pointsTeam1.":".$ergebnis->pointsTeam2.")";
                        }
                    }
                    if ($endergebnis == '') {
                        $endergebnis = '0:0';
                    }
                    $tootip_text .= $endergebnis . $halbzeitergebnis;

                    $goals = '';
                    foreach ($partie->goals as $goal) {
                        if ($goal->goalGetterName) {
                            if ($goal->matchMinute) {
                                $goals .= '<b>' . $goal->scoreTeam1 . ':' . $goal->scoreTeam2 . '</b>&nbsp;&nbsp;' . $goal->goalGetterName . ' (' . $goal->matchMinute . '.)<br>';
                            } else {
                                $goals .= '<b>' . $goal->scoreTeam1 . ':' . $goal->scoreTeam2 . '</b>&nbsp;&nbsp;' . $goal->goalGetterName . '<br>';
                            }
                        }
                    }
                }
                if (isset($partie->matchIsFinished)) {
                    if (!$partie->matchIsFinished && $alle_ergebnisse[0] instanceof stdClass) {
                        $tootip_text .= "</font>";
                    }
                }

                $tootip_text .= "</nobr>";

                if ($goals <> '') {
                    $table .= JHtml::_('tooltip', $goals, 'Tore', '', $tootip_text);
                } else {
                    $table .= $tootip_text;
                }

                $table .= "</td>\r\n";
            } elseif ($jparams->get('longnames') == '1') {
                $table .= "<td>-:- (-:-)</td>\r\n";
            }
            $table .= "</tr>\r\n";
        }

        $table .= "</table>\r\n";

        return $table;
    }
}
